utils/zerovec/design_doc.md

```diff
@@ -0,0 +1,417 @@
+    println!("{:?}", ch);
+}
+```
+
+### Vectors of variable-size types
+
+Vectors of variable-size types work via [`VarZeroVec<'a, T>`][`VarZeroVec`], where `'a` is the lifetime of the borrowed data. `T` is typically an unsized type like `str`, `[u8]`, `ZeroSlice`, or `VarZeroSlice` (more on this type later).
+
+```rust
+use zerovec::{VarZeroVec, ZeroSlice};
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct Foo<'a> {
+    #[serde(borrow)]
+    strings: VarZeroVec<'a, str>, // behaves like Vec<String>
+    #[serde(borrow)]
+    nested_numbers: VarZeroVec<'a, ZeroSlice<u32>>, // Behaves like Vec<Vec<u32>>
+}
+
+let deserialized: Foo = postcard::from_bytes(/* some bytes */);
+
+for s in deserialized.strings.iter() {
+    println!("{:?}", s);
+}
+
+```
+
+The main differences from `Vec` are that:
+
+ - Only types which implement `VarULE` are allowed inside `VarZeroVec<T>`. More on this trait later.
+ - `VarZeroVec<'a, T>` dereferences to [`VarZeroSlice<T>`][`VarZeroSlice`]. This is kind of like `[T]`, except since `T` is not `Sized`, types like `[str]` do not actually exist in Rust.
+
+
+The `Owned` variant contains a `VarZeroVecOwned` (which can also be accessed by `VarZeroVec::make_mut()`), which allows for regular vector mutation operations like inserting and removing. These operations are somewhat expensive since under the hood it is mutating a single flat buffer.
+
+`VarZeroSlice<T>` and `ZeroSlice<T>` are both types that allow you to arbitrarily nest vectors, so a `Vec<Vec<Vec<Vec<u32>>>>` in zero-copy form would be `VarZeroVec<'a, VarZeroSlice<VarZeroSlice<ZeroSlice<u32>>>>`. They're also useful if you want to zero-copy deserialize in cases where you know you do not need owned variants.
+
+#### Internal representation of `VarZeroVec`
+
+Internally, a `VarZeroVec<'a, T>`'s buffer is laid out as follows:
+
+ - 4 bytes specifying the "length" of the vector N
+ - 4N bytes of "indices" specifying the ending index of each item in the "data" slice
+ - However many bytes are needed to represent the actual "data"
+ 
+This allows for packing variable-size data together into a single flat buffer that can hand out references. Users largely need not care about this representation, however it should give an idea of how expensive mutation operations may be.
+
+### Maps
+
+Maps are handled by [`ZeroMap<'a, K, V>`][`ZeroMap`], where `K` and `V` can both be either an `AsULE` or `VarULE` type. Internally, it works by selecting the appropriate vector type for the keys and for the values (`ZeroVec` or `VarZeroVec`), and setting up a sorted array that uses binary search for lookup. In its essence, it is a zero-copy version of [`LiteMap`].
+
+Operations have the same quirks that the underlying array types have, so for example `.get()` with a `V = str` will return an `&str`, but with `V = u32` it will return `u32`. Basically, `Sized` keys and values are copied in and out, and dynamically sized ones are returned as references.
```

Comment:
`.get()` returns references to ULE types.  `.get_copied()` is the special method that turns `&u32::ULE` into `u32`

Reply:
Good point